[CMake] Fix using precompiled headers with ccache

mrkajetanp opened the pull request:
Using precompiled headers with ccache requires special accommodations.
Add the required CCACHE option and clang compiler flag to CMake.

mrkajetanp committed:
[CMake] Fix using precompiled headers with ccache

Using precompiled headers with ccache requires special accommodations.
Add the required CCACHE option and clang compiler flag to CMake.

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

mrkajetanp commented:
@tblah @DavidTruby @mgorny @Meinersbur for review.

I'm not sure whether this needs special handling for Windows w.r.t. that `Ccache configuration through CMake is not supported on Windows. Please use environment variables.` line or not.

DavidTruby commented:
In my experience `ccache` is _essentially_ non-functional on Windows. Other people might have had more success with it than me though.

Meinersbur commented:
@mrkajetanp Did you check whether with these changes, ccache does cache those compilations?

I am using ccache under Windows and works fine. I could not reproduce the error in https://lab.llvm.org/staging/#/builders/206/builds/174.

mrkajetanp commented:
> @mrkajetanp Did you check whether with these changes, ccache does cache those compilations?

A fresh ccache log is showing cache hits for the precompiled headers with this patch just as without it, yes.

Meinersbur commented on llvm/CMakeLists.txt:
```diff
@@ -291,7 +291,7 @@ if(LLVM_CCACHE_BUILD)
     else()
       if(LLVM_CCACHE_MAXSIZE OR LLVM_CCACHE_DIR OR
-         NOT LLVM_CCACHE_PARAMS MATCHES "CCACHE_CPP2=yes CCACHE_HASHDIR=yes")
+         NOT LLVM_CCACHE_PARAMS MATCHES "CCACHE_CPP2=yes CCACHE_HASHDIR=yes CCACHE_SLOPPINESS=pch_defines,time_macros")
```
If your Windows question was for this, I it will cause no cache hits because nothing sets `CCACHE_SLOPPINESS=pch_defines,time_macros`.

Workarounds could be 
1. `env` command from MSYS
2. `set` command from Windows itself (only a single variable per invocation and does not support specifying a command on the same line but can be  chained with `&&`): https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/set_1
3. emit a wrapper around `ccache` that sets those enviornment variables
4. test & abort if ccache does not use the required sloppiness option

mrkajetanp replied:
As written this will end up with compilation commands like
`"/usr/local/bin/cmake -E env CCACHE_CPP2=yes CCACHE_HASHDIR=yes CCACHE_SLOPPINESS=pch_defines,time_macros -- /usr/bin/ccache" /usr/bin/clang++ -DGTEST_H(..)`
and fail with
`/bin/sh: 1: /usr/local/bin/cmake -E env CCACHE_CPP2=yes CCACHE_HASHDIR=yes CCACHE_SLOPPINESS=pch_defines,time_macros -- /usr/bin/ccache: not found`
due to `launcher` getting put in quote marks. Removing the quotemarks from the last two lines in the patch above does not seem to make a difference, is there some way we could get around it?

mrkajetanp replied:
Ah ok, we just need to be careful with the quoting. This works:
```
+    set(launcher ${CMAKE_COMMAND} -E env ${launcher_env} ${CCACHE_PROGRAM})
+    set(CMAKE_C_COMPILER_LAUNCHER ${launcher})
+    set(CMAKE_CXX_COMPILER_LAUNCHER ${launcher})
```

mrkajetanp replied:
See the new commit for a working solution.

mrkajetanp commented:
Just to save people from clicking through the related PRs, here's the ccache documentation I based this off of:
https://ccache.dev/manual/latest.html#_precompiled_headers

mgorny commented on flang/CMakeLists.txt:
```diff
@@ -446,6 +446,11 @@ if (APPLE)
 
+# Clang requires this flag in order for precompiled headers to work with ccache.
+if (CMAKE_CXX_COMPILER_ID MATCHES Clang)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Xclang -fno-pch-timestamp")
```
For the record, I think it's better to have it globally. At least in Gentoo we have ccache enabled at package manager level, so having this globally would help us — and I don't think there's any harm in doing that, is there?

Meinersbur replied:
On package-manager level, when setting `CMAKE_*_LAUNCHER`, can't you just also set `CMAKE_*_FLAGS=-Xclang -fno-pch-timestamp`? That doesn't need support from the build system of each package.

mgorny replied:
Sure — but the point is, I wouldn't even have realized that I need to set something special. Not a big deal for me, just pointing out that PCHs aren't that common and people don't have awareness that clang requires special flags for, well, reproducible builds.

tblah replied:
Another thought, how recent is clang's PCH support? Is -fno-pch-timestamp supported on all versions of clang supported by LLVM?

mrkajetanp replied:
The option was added in 2016 so it should be available from clang 4.0 onwards.
Original commit: 533a893fa126d8a9662d836ad28f51afa1d2e74a

Meinersbur replied:
If the users sets CMAKE_CXX_COMPILER_LAUNCHER without using `LLVM_CCACHE_BUILD`, then they are themsolves responsible for making the accomodations. They will also need set the sloppiness options or they will not gain anything.

However, I am not complety opposed to add it unconditionally, I just think generally defaults should be used if there is not reason to diverge from them.

mrkajetanp replied:
Yes, I suppose you're right. These sorts of exotic setups will come out broken in some ways no matter what we do here. I think I'm leaning towards the same position as @mgorny, but not particularly strongly. If there's wider agreement that we should be checking for LLVM_CCACHE_BUILD I don't mind adding the check all that much. @petrhosek any thoughts on this?

mrkajetanp replied:
My understanding is that clang will look at the file size and the timestamp in order to work out whether it needs to be recompiled or not. Without the timestamp, it's just left with having to rely on whether the size has changed or not. Looking at [the original commit message](https://github.com/llvm/llvm-project/commit/533a893fa126d8a9662d836ad28f51afa1d2e74a) it seems as if the flag was added after people realised the pch timestamps were breaking distributed build systems. So maybe it's not the default because it was done after the fact. Though I don't have any special insight there outside of just inferring things.

mrkajetanp committed:
[CMake] Configure ccache using command line options

Since ccache 4.8, it is possible to pass configuration options directly
to ccache on the command line as opposed to through environment
variables.
This is the intended way for CMake to configure ccache, as described on
the GitHub wiki for the project:
https://github.com/ccache/ccache/wiki/CMake

Rework the way ccache is configured by LLVM accordingly.

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

mrkajetanp committed:
[flang] [CMake] Set compiler flags for ccache+pch

Reference ccache documentation:
https://ccache.dev/manual/latest.html#_precompiled_headers

For clang, only pass -Xclang -fno-pch-timestamp if precompiled headers are enabled
in CMake.

For gcc, pass -fpch-preprocess in the same scenario.

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

mrkajetanp committed:
[CMake] Add a workaround for pre-4.8 ccache configuration

With ccache only supporting passing arguments on the command line since
version 4.8, a workaround is needed for systems with older versions.
Whenever possible, we want to use the new way of configuring ccache
because it will ensure uniform behaviour across platforms in a way that
just setting environment variables might not.

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

DavidTruby reviewed (approved):
I've done a little investigation and the reason this flag isn't default anyway when using .pch files is essentially to work around/protect against dodgy build systems that won't rebuild the pch properly in all cases when the .h has changed. So clang will error if the .pch hasn't been updated properly.

We can rely on CMake handling this correctly; the target has a dependency on the .pch and the .pch has a dependency on the .h so there will be no issue here. So I think it's fine to just add this flag in every case even when ccache/other caching software isn't being used.

LGTM on this basis but please wait for the other reviewers to comment to check that they agree. Thanks for the fix!

mrkajetanp commented:
@Meinersbur Do you think this is okay to merge now?

Meinersbur reviewed (approved):
LGTM, thank you

Meinersbur reviewed (approved):
LGTM, thank you

petrhosek commented on flang/CMakeLists.txt:
```diff
@@ -436,6 +436,10 @@ if (LLVM_COMPILER_IS_GCC_COMPATIBLE)
    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -fno-semantic-interposition")
   endif()
 
+  # gcc requires this flag in order for precompiled headers to work with ccache
+  if (NOT CMAKE_CXX_COMPILER_ID MATCHES Clang AND NOT CMAKE_DISABLE_PRECOMPILE_HEADERS)
```
Since this flag is required by GCC, can we test that the compiler is GCC?

```suggestion
  # GCC requires this flag in order for precompiled headers to work with ccache
  if (CMAKE_CXX_COMPILER_ID MATCHES GCC AND NOT CMAKE_DISABLE_PRECOMPILE_HEADERS)
```

mrkajetanp replied:
I was following what the rest of the file was doing in matching on "not clang" in this context, but yes of course. Done.

petrhosek commented on llvm/CMakeLists.txt:
```diff
@@ -275,30 +275,47 @@ set(LLVM_CCACHE_BUILD OFF CACHE BOOL "Set to ON for a ccache enabled build")
+    # ccache --version example output: "ccache version 4.9.1\n(..)"
+    execute_process(COMMAND ${CCACHE_PROGRAM} --version OUTPUT_VARIABLE CCACHE_VERSION_STR)
+    # Strip the prefix before the version number, taking some extra characters
+    string(SUBSTRING "${CCACHE_VERSION_STR}" 15 10 CCACHE_VERSION_STR)
```
Since the version has a fixed format, could we use regular expression instead?

mrkajetanp replied:
Ah indeed that's a much better way to do it, thanks!

mrkajetanp committed:
[CMake] Use regex to extract ccache version

flang - Explicitly match gcc to add gcc-specific flag.

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

mgorny reviewed (approved):
It seems to work fine with GCC, at least for me.

tblah commented:
@petrhosek are you happy for this to be merged?

mrkajetanp commented:
There are bug reports like the one in #134726 waiting for this to be merged, so given there are 4 approvals I'll just go ahead and merge this. If something needs to be changed or if there are other objections, we can fix or revert later.

mrkajetanp commented:
Reverted for now on account of the Windows issues. I think there's a difference with how CMake expands `"${CCACHE_PROGRAM};${launcher_params}"` on some platforms? On Linux it expands to `ccache <params>` but apparently on this Windows build bot the whole thing was ending up quoted:
```
"C:\ninja\ccache.exe run_second_cpp=true hash_dir=true sloppiness=pch_defines,time_macros" C:\PROGRA~1
```

mrkajetanp commented:
> And can that change be split out from the functional change in passing more options wrt PCH

Certainly. Originally this was just a change to add the required flags for PCH, but in a review comment I was asked to also refactor the way we configure ccache while at it because it is currently using some antipatterns. It does make more sense to do it separately and the other way around, I'll do that.
